overlord/snapstate/reboot.go

```diff
@@ -123,6 +229,241 @@ func bootBaseSnapType(byTypeTss map[snap.Type]*state.TaskSet) snap.Type {
+	if ts := byTypeTss[snap.TypeSnapd]; ts != nil {
+		// Snapd does not need to be part of the transaction, as
+		// it's not really necessary to undo snapd should base/kernel/gadget
+		// fail.
+		const transactional = false
+		const split = false
+		if err := chainEssentialTs(ts, snap.TypeSnapd, transactional, split); err != nil {
+			return err
+		}
+	}
+
+	// Is gadget present? Then we cannot do single-reboot for now
+	// TODO: enable this
+	gadgetTs := byTypeTss[snap.TypeGadget]
+	bootSnapType := bootBaseSnapType(byTypeTss)
+
+	// Then we link in the boot-base, to run after snapd, it could run in it's
+	// entirety before a reboot, as we expect boot-bases to be 'simple' and not
+	// have any hooks.
+	if ts := byTypeTss[bootSnapType]; ts != nil {
+		const transactional = true
+		// We can only do split on this if there is no gadget.
+		split := (gadgetTs == nil || enforcedSingleRebootForGadgetKernelBase)
+		if err := chainEssentialTs(ts, bootSnapType, transactional, split); err != nil {
+			return err
+		}
+	}
+
+	// Next we link in the gadget, and it needs to be part of the transaction
+	// so it will be undone in the event of failures.
+	if gadgetTs != nil {
+		const transactional = true
+		const split = false
+		if err := chainEssentialTs(gadgetTs, snap.TypeGadget, transactional, split); err != nil {
+			return err
+		}
+	}
+
+	// Then we link in the kernel, it needs to run latest, but before other bases and apps.
+	if ts := byTypeTss[snap.TypeKernel]; ts != nil {
+		const transactional = true
+		// We can only do split on this if there is no gadget.
+		split := (gadgetTs == nil || enforcedSingleRebootForGadgetKernelBase)
+		if err := chainEssentialTs(ts, snap.TypeKernel, transactional, split); err != nil {
+			return err
+		}
+	}
+
+	// Now link in all the after task-sets that have been split, which should run
+	// post-reboot.
+	// XXX: Review and validate that this order is correct once we put in gadget
+	for _, o := range essentialSnapsRestartOrder {
+		if afterTss[o] == nil {
+			continue
+		}
+		const transactional = false
+		const split = false
+		if err := chainEssentialTs(afterTss[o], o, transactional, split); err != nil {
+			return err
+		}
+	}
+
+	// Ensure restart boundaries are set, for the task-sets that have been
+	// split, we ensure that boundaries are set *only* for the last of those, to allow
+	// them all to run before the reboot.
+	if len(beforeTss) > 0 {
+		if err := arrangeSingleRebootForSplitTaskSets(beforeTss); err != nil {
+			return err
+		}
+	}
+
+	// For the task-sets that have not been split, they must have boundaries set for
+	// each of them. Reuse the restart order list here, so we go through the correct
+	// list of snap types that needs restart boundaries set.
+	for _, o := range essentialSnapsRestartOrder {
+		// Make sure that the core snap is actually the boot-base
+		if o == snap.TypeOS && bootBase != "core" {
+			continue
+		}
+		// If it was split, it has already been handled
```

Comment:
Now this comment is a bit confusing as it does not match the statement below, should be "if it was not split, we need to set the boundaries now" or something similar

Reply:
Good catch! Missed this while changing the statement, clarified this comment